esmvalcore/preprocessor/_multimodel.py

```diff
@@ -201,158 +185,88 @@ def _put_in_cube(template_cube, cube_data, statistic, t_axis):
 
-    Simple cube data slicer on indices
-    of common time-data elements.
-    """
-    time_pts = [t for t in cube.coord('time').points]
-    converted_t = _datetime_to_int_days(cube)
-    idxs = sorted([
-        time_pts.index(ii) for ii, jj in zip(time_pts, converted_t)
-        if t_1 <= jj <= t_2
-    ])
-    return [idxs[0], idxs[-1]]
-
-
-def _monthly_t(cubes):
-    """Rearrange time points for monthly data."""
-    # get original cubes tpoints
-    days = {day for cube in cubes for day in _datetime_to_int_days(cube)}
-    return sorted(days)
-
-
-def _full_time_slice(cubes, ndat, indices, ndatarr, t_idx):
-    """Construct a contiguous collection over time."""
-    for idx_cube, cube in enumerate(cubes):
-        # reset mask
-        ndat.mask = True
-        ndat[indices[idx_cube]] = cube.data
-        if np.ma.is_masked(cube.data):
-            ndat.mask[indices[idx_cube]] = cube.data.mask
+    for cube in cubes:
+        # Extract date info from cube
+        years = [cell.point.year for cell in cube.coord('time').cells()]
+        months = [cell.point.month for cell in cube.coord('time').cells()]
+
+        # Reconstruct default calendar
+        if 0 not in np.diff(years):
+            # yearly data
+            dates = [datetime(year, 7, 1) for year in years]
+
+        elif 0 not in np.diff(months):
+            # monthly data
+            dates = [
+                datetime(year, month, 15)
+                for year, month in zip(years, months)
+            ]
         else:
-            ndat.mask[indices[idx_cube]] = False
-        ndatarr[idx_cube] = ndat[t_idx]
-
-    # return time slice
-    return ndatarr
-
-
-def _assemble_overlap_data(cubes, interval, statistic):
-    """Get statistical data in iris cubes for OVERLAP."""
-    start, stop = interval
-    sl_1, sl_2 = _slice_cube(cubes[0], start, stop)
-    stats_dats = np.ma.zeros(cubes[0].data[sl_1:sl_2 + 1].shape)
+            # (sub)daily data
+            raise ValueError("Multimodel only supports yearly or monthly data")
 
-    # keep this outside the following loop
-    # this speeds up the code by a factor of 15
-    indices = [_slice_cube(cube, start, stop) for cube in cubes]
-
-    for i in range(stats_dats.shape[0]):
-        time_data = [
-            cube.data[indx[0]:indx[1] + 1][i]
-            for cube, indx in zip(cubes, indices)
-        ]
-        stats_dats[i] = _compute_statistic(time_data, statistic)
-    stats_cube = _put_in_cube(cubes[0][sl_1:sl_2 + 1],
-                              stats_dats,
-                              statistic,
-                              t_axis=None)
-    return stats_cube
+        # Update the cubes' time coordinate (both point values and the units!)
+        cube.coord('time').points = [t_unit.date2num(date) for date in dates]
+        cube.coord('time').units = t_unit
+        cube.coord('time').bounds = None
+        cube.coord('time').guess_bounds()
```

Comment:
This doesn't work if there is only 1 time point. Related question: why is a time coordinate needed to compute a multimodel statistic?

Reply:
what do you mean why is a t-coord needed for a multimodel stat? :grin: you compute a time-dependent statistic `S(t) = \sum_i^N S_i(t) / N`

Reply:
Created an issue for this: #738

Reply:
Let's solve this in a separate PR.